### PR TITLE
feat(metadata): adding ControllerTimeoutDecorator to prevent overloading of the main Controller

### PR DESCRIPTION
**Pull request by DaniilRoman**
PR solves the issue https://github.com/AutoMQ/automq/issues/1446
Added a unit test to test a new functionality 

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by superhx** [comment]
Kafka assumes that requests will be processed sequentially. Changing the queue to a PriorityQueue may introduce unknown risks.

https://github.com/AutoMQ/automq/issues/1446 The core purpose is to solve how to handle scenarios where the Controller (EventLoop / CPU / GC) is overloaded, ensuring that the Controller does not mistakenly consider all Brokers as Fenced, triggering massive partition migrations.

File: metadata/src/main/java/org/apache/kafka/controller/ControllerTimeoutDecorator.java [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+ */
+public class ControllerTimeoutDecorator implements Controller {
+
+    private final Controller controller;
+
+    private final int defaultTimeoutInMs;
+
+    public ControllerTimeoutDecorator(Controller controller) {
+        this.controller = controller;
+        this.defaultTimeoutInMs = 3000;
```
Comment (DaniilRoman) [inline metadata/src/main/java/org/apache/kafka/controller/ControllerTimeoutDecorator.java:106]: I assume you have the necessary statistic data to be able to correct the default timeout

File: metadata/src/main/java/org/apache/kafka/controller/ControllerTimeoutDecorator.java [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+    public CompletableFuture<Integer> getNextNodeId(
+            ControllerRequestContext context,
+            GetNextNodeIdRequestData request
+    ) {
+        return controller.getNextNodeId(context, request);
+    }
+
+    @Override
+    public CompletableFuture<Void> checkS3ObjectsLifecycle(ControllerRequestContext context) {
+        return controller.checkS3ObjectsLifecycle(context).orTimeout(defaultTimeoutInMs, TimeUnit.MILLISECONDS);
```
Comment (DaniilRoman) [inline metadata/src/main/java/org/apache/kafka/controller/ControllerTimeoutDecorator.java:287]: Added timeout only for 2 new methods and that are running by schedule

File: core/src/main/scala/kafka/server/ControllerServer.scala [hunk 3]
```diff
@@ -581,9 +582,6 @@ class ControllerServer(
   }
 
   // AutoMQ for Kafka inject start
-  protected def quorumControllerExtension(quorumController: QuorumController): QuorumControllerExtension = {
-    QuorumControllerExtension.NOOP
-  }
```
Comment (DaniilRoman) [inline core/src/main/scala/kafka/server/ControllerServer.scala:586]: It wasn't used anywhere

File: core/src/main/scala/kafka/server/ControllerServer.scala [hunk 4]
```diff
@@ -117,7 +116,7 @@ class ControllerServer(
   var createTopicPolicy: Option[CreateTopicPolicy] = None
   var alterConfigPolicy: Option[AlterConfigPolicy] = None
   @volatile var quorumControllerMetrics: QuorumControllerMetrics = _
-  var controller: QuorumController = _
+  var controller: Controller = _
```
Comment (DaniilRoman) [inline core/src/main/scala/kafka/server/ControllerServer.scala:119]: Initially there was Controller as well

**Comment by DaniilRoman** [comment]
Thank you @superhx for the clarifications 🙏  Initially I was looking at the wrong direction. I found [that docs](https://docs.automq.com/automq/what-is-automq/compatibility-with-apache-kafka#apache-kafka-compatibility-overview) where was said that AutoMQ doesn't even use old ZK mode and instead AutoMQ always runs in KRaft mode

File: metadata/src/main/java/org/apache/kafka/controller/ControllerTimeoutDecorator.java (hunk 2 above)
Comment (superhx) [reply metadata/src/main/java/org/apache/kafka/controller/ControllerTimeoutDecorator.java:287]: The timeout only complete the CompletableFuture with TimeoutException, the `#checkS3ObjectsLifecycle` is still running in the EventLoop.

**Comment by superhx** [comment]
Do you mind if I solve this issue in this way? https://github.com/AutoMQ/automq/issues/1446

**Comment by superhx** [comment]
https://github.com/AutoMQ/automq/issues/1446 resolved

**Pull request closed without merging**